### ISSUE TITLE
spent task cleanup can be broken by some cylc-6 graphs

### ISSUE THREAD

**Issue by hjoliver** [cylc/cylc#1392]
Spent tasks in this suite will never be cleaned up, because of our current reliance on a "clean-up cuttoff" cycle point for each task:

```cfg
[scheduling]
    initial cycle point = 1
    cycling mode = integer
    [[dependencies]]
        [[[R//P1]]]
            graph = foo
        [[[R/+P1/P1]]]
            graph = foo[^] => foo
```

The dependence on `foo[^]` here is useless, but that's beside the point.

The proper solution to this (updated from bullet point 2 in this old issue comment: https://github.com/cylc/cylc/issues/108#issuecomment-8654425) is probably to use the run-db to satisfy the prerequisites of other tasks, then we can get rid of any finished task immediately (or perhaps keep one instance of each in the pool for monitoring purposes and to allow easy retriggering).  The only question is - @arjclark - would the extra db interrogations have performance implications?

#1773 is marked as superseded by this. It also mentions the case `foo_cold[^] => foo`. In particular:
* For a large enough suite, with a large enough number of cycles, this daemon computation time can be very significant.
* Can we do anything to jump over the intermediate graph nodes / cycle points?

See also #2329.

**Comment by arjclark** [comment]
I guess it would depend on what queries you make, how often and when.

Potentially, one could query the db for out of scope task events when a task object is created (much like submit number is now) and set the dependencies accordingly (N.B. only out of scope ones should be queried to minimize load increases). I'm unsure what the implication of this would be in terms of db loading so would need careful testing under heavy usage, e.g. with the loadtest suite.

**Comment by hjoliver** [comment]
[meeting]
 * this is a very unusual suite structure, with an easy workaround - use a different name for the first task, and it will become irrelevant after #987 so ... leave open in some-day, mention in FAQ

**Comment by matthewrmshin** [comment]
(Updated description.)

**Comment by hjoliver** [comment]
Closing as superseded by #2329.
